File: Sources/ThesisSkipApp/utils/AppLogger.swift
//
//  AppLogger.swift
//  MADiOS
//
//  Created by Bram Rampelberg on 04/01/2025.
//  Copyright © 2025 HOGENT. All rights reserved.
//

import Foundation
import os

final class AppLogger {
    static private let logger = Logger(subsystem: "be.hogent.ThesisSkipApp", category: "ThesisSkipApp")
    
    static private var timestamp: String {
        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "yyyy-MM-dd HH:mm:ss"
        return dateFormatter.string(from: Date())
    }
    
#if SKIP
    static func info(_ message: Any){
        log(level: Level.info, message: message)
    }
    
    static func error(_ message: Any){
        log(level: Level.error, message: message)
    }
    
    static func debug(_ message: Any){
        log(level: Level.debug, message: message)
    }
    
    static func warning(_ message: Any){
        log(level: Level.warning, message: message)
    }
    
    static private func log(level: Level, message: Any){

        let formattedMessage = "\(level) : [\(timestamp)] \(message)"
        
        switch level{
        case .info: logger.info("\(formattedMessage)")
        case .error: logger.error("\(formattedMessage)")
        case .debug: logger.debug("\(formattedMessage)")
        case .warning: logger.warning("\(formattedMessage)")
        }
    }
#else
    static func info(_ message: Any, file: String = #file, function: String = #function, line: Int = #line){
        log(level: .info, message: message, file: file, function: function, line: line)
    }
    
    static func error(_ message: Any, file: String = #file, function: String = #function, line: Int = #line){
        log(level: .error, message: message, file: file, function: function, line: line)
    }
    
    static func debug(_ message: Any, file: String = #file, function: String = #function, line: Int = #line){
        log(level: .debug, message: message, file: file, function: function, line: line)
    }
    
    static func warning(_ message: Any, file: String = #file, function: String = #function, line: Int = #line){
        log(level: .warning, message: message, file: file, function: function, line: line)
    }
    
    static private func log(level: Level, message: Any, file: String, function: String, line: Int){
        let fileName = (file as NSString).lastPathComponent

        let formattedMessage = "\(level) : [\(timestamp)] \(message) [\(fileName):\(line) \(function)]"
        
        switch level{
        case .info: logger.info("\(formattedMessage)")
        case .error: logger.error("\(formattedMessage)")
        case .debug: logger.debug("\(formattedMessage)")
        case .warning: logger.warning("\(formattedMessage)")
        }
    }
#endif
    
    private enum Level {
        case info
        case error
        case debug
        case warning
    }
    
    private init(){}
}

File: .build/plugins/outputs/thesis-skip-app/ThesisSkipApp/destination/skipstone/ThesisSkipApp/src/main/kotlin/thesis/skip/app/AppLogger.kt
//
//  AppLogger.swift
//  MADiOS
//
//  Created by Bram Rampelberg on 04/01/2025.
//  Copyright © 2025 HOGENT. All rights reserved.
//

package thesis.skip.app

import skip.lib.*

import skip.foundation.*

internal class AppLogger {


    private enum class Level {
        info,
        error,
        debug,
        warning;
    }

    private constructor() {
    }

    companion object {
        private val logger = SkipLogger(subsystem = "be.hogent.ThesisSkipApp", category = "ThesisSkipApp")

        private val timestamp: String
            get() {
                val dateFormatter = DateFormatter()
                dateFormatter.dateFormat = "yyyy-MM-dd HH:mm:ss"
                return dateFormatter.string(from = Date())
            }
        internal fun info(message: Any): Unit = log(level = Level.info, message = message)

        internal fun error(message: Any): Unit = log(level = Level.error, message = message)

        internal fun debug(message: Any): Unit = log(level = Level.debug, message = message)

        internal fun warning(message: Any): Unit = log(level = Level.warning, message = message)

        private fun log(level: AppLogger.Level, message: Any) {

            val formattedMessage = "${level} : [${timestamp}] ${message}"

            when (level) {
                AppLogger.Level.info -> logger.info("${formattedMessage}")
                AppLogger.Level.error -> logger.error("${formattedMessage}")
                AppLogger.Level.debug -> logger.debug("${formattedMessage}")
                AppLogger.Level.warning -> logger.warning("${formattedMessage}")
            }
        }
    }
}
